src/integration/kotlin/pl/edu/agh/gem/integration/controller/ExpenseControllerIT.kt

```diff
@@ -44,235 +50,303 @@ import pl.edu.agh.gem.internal.validation.ValidationMessage.USER_NOT_PARTICIPANT
+            val createExpenseRequest = createExpenseCreationRequest(baseCurrency = CURRENCY_1, targetCurrency = CURRENCY_2)
+            stubGroupManagerMembers(groupMembers, GROUP_ID)
+            stubGroupManagerOptions(createGroupOptionsResponse(currencies = createCurrencies(CURRENCY_1)), GROUP_ID)
+            stubCurrencyManagerAvailableCurrencies(createCurrenciesResponse(CURRENCY_1, CURRENCY_2))
+            stubCurrencyManagerExchangeRate(ExchangeRateResponse(EXCHANGE_RATE_VALUE), CURRENCY_1, CURRENCY_2, Instant.ofEpochSecond(0L))
+
+            // when
+            val response = service.createExpense(createExpenseRequest, createGemUser(USER_ID), GROUP_ID)
+
+            // then
+            response shouldHaveHttpStatus BAD_REQUEST
+            response shouldHaveValidatorError TARGET_CURRENCY_NOT_IN_GROUP_CURRENCIES
+        }
+
+        should("return validator exception cause BASE_CURRENCY_NOT_AVAILABLE") {
+            // given
+            val groupMembers = GroupMembersResponse(listOf(GroupMemberResponse(USER_ID), GroupMemberResponse(OTHER_USER_ID)))
+            val createExpenseRequest = createExpenseCreationRequest(baseCurrency = CURRENCY_1, targetCurrency = CURRENCY_2)
+            stubGroupManagerMembers(groupMembers, GROUP_ID)
+            stubGroupManagerOptions(createGroupOptionsResponse(currencies = createCurrencies(CURRENCY_2)), GROUP_ID)
+            stubCurrencyManagerAvailableCurrencies(createCurrenciesResponse(CURRENCY_2))
+            stubCurrencyManagerExchangeRate(ExchangeRateResponse(EXCHANGE_RATE_VALUE), CURRENCY_1, CURRENCY_2, Instant.ofEpochSecond(0L))
+
+            // when
+            val response = service.createExpense(createExpenseRequest, createGemUser(USER_ID), GROUP_ID)
+
+            // then
+            response shouldHaveHttpStatus BAD_REQUEST
+            response shouldHaveValidatorError BASE_CURRENCY_NOT_AVAILABLE
+        }
+
+        should("get expense") {
+            // given
+            val groupMembers = GroupMembersResponse(listOf(GroupMemberResponse(USER_ID), GroupMemberResponse(OTHER_USER_ID)))
+            stubGroupManagerMembers(groupMembers, GROUP_ID)
+            val expense = createExpense(expenseParticipants = listOf(createExpenseParticipant()))
+            expenseRepository.create(expense)
+
+            // when
+            val response = service.getExpense(createGemUser(USER_ID), expense.id, GROUP_ID)
+
+            // then
+            response shouldHaveHttpStatus OK
+            response.shouldBody<ExpenseResponse> {
+                creatorId shouldBe expense.creatorId
+                cost shouldBe expense.cost
+                baseCurrency shouldBe expense.baseCurrency
+                targetCurrency shouldBe expense.targetCurrency
+                exchangeRate shouldBe expense.exchangeRate?.value
+                createdAt shouldBe expense.createdAt.truncatedTo(MILLIS)
```

Comment:
i dont sure if we should tructedthat to milims , because we have to difrent values in each side,IMHO check if we have not nullvalues or mock clock for that to have persistance